✨ feat: support server db mode with Postgres / Drizzle ORM / tRPC

arvinxx opened the pull request:
#### 💻 变更类型 | Change Type

<!-- For change type, change [ ] to [x]. -->

- [x] ✨ feat
- [ ] 🐛 fix
- [ ] ♻️ refactor
- [ ] 💄 style
- [ ] 🔨 chore
- [ ] ⚡️ perf
- [ ] 📝 docs

#### 🔀 变更说明 | Description of Change

### 整体进展

LobeChat 服务端 DB 版本功能实现，目前完成度为 100% 。

> \[!Note]
> 
> 由于之前并没有写过 SQL 和 Postgres ，所以可能整体服务端实现上仍然存在瑕疵，也可能存在 critical 的问题， 欢迎大家点评指正。

### 技术栈

- **DB**：Postgres ，默认采用 [Neon](https://neon.tech)；
- **ORM**: Dirzzle ORM
- **Auth**: Clerk 
- **server router**: tRPC

由于基于 Drizzle ORM， 非常期待看到未来拓展到其他 Postgres DB、甚至 SQLite、MySQL 。 以及有待 @cy948 帮忙实现 next-auth 相关的 db 鉴权方案。

### 待解决工程问题

工程链路上还有两个问题待解决：

- [x] 1. 数据库无感 migration ： https://github.com/lobehub/lobe-chat/discussions/1851#discussioncomment-9662082 ；
- [x] 2. 带有真实数据库的 ci/cd workflow ： https://github.com/lobehub/lobe-chat/discussions/1851#discussioncomment-9662165

<!-- Thank you for your Pull Request. Please provide a description above. -->

### 服务端相关的配置

```env
# Postgres DB Url
DATABASE_URL=

# 指定 service mode 为 server 
NEXT_PUBLIC_SERVICE_MODE=server

# Clerk 相关配置
NEXT_PUBLIC_CLERK_PUBLISHABLE_KEY=
CLERK_SECRET_KEY=
CLERK_WEBHOOK_SECRET=

# S3 相关配置（上传图片必须）
NEXT_PUBLIC_S3_DOMAIN=

S3_ACCESS_KEY_ID=
S3_SECRET_ACCESS_KEY=
S3_ENDPOINT=
S3_BUCKET=
```

#### 📝 补充信息 | Additional Information

- https://github.com/lobehub/lobe-chat/discussions/1851
<!-- Add any other context about the Pull Request here. -->

- close #1861
- close #1573 
- close #1354 
- close #1547

cfwl1111 commented:
目前是可以配置服务端数据库了吗

arvinxx commented:
> 目前是可以配置服务端数据库了吗

@cfwl1111 等这个 PR 合并后，就可以了

lee88688 commented:
图片的保存可以考虑支持直接存储在PostgreSQL数据库中。自部署还需要单独起一个S3兼容的后端服务管理起来更复杂，特别是大部分自部署用户的使用强度不是那么高，使用数据库完全没问题。

arvinxx commented:
> 图片的保存可以考虑支持直接存储在PostgreSQL数据库中。自部署还需要单独起一个S3兼容的后端服务管理起来更复杂，特别是大部分自部署用户的使用强度不是那么高，使用数据库完全没问题。

不建议，后面还要支持各种文件上传的。如果不想承担这个复杂度，要么可以用本地部署的版本，不需要s3，或者用未来我们host的cloud 版

arvinxx committed:
✨ feat: support server db with Postgres and drizzle ORM

semantic-release-bot committed:
:bookmark: chore(release): v0.162.25 [skip ci]

### [Version&nbsp;0.162.25](https://github.com/lobehub/lobe-chat/compare/v0.162.24...v0.162.25)
<sup>Released on **2024-06-16**</sup>

#### 🐛 Bug Fixes

- **misc**: Fix issues for client fetch.

<br/>

<details>
<summary><kbd>Improvements and Fixes</kbd></summary>

#### What's fixed

* **misc**: Fix issues for client fetch, closes [#2753](https://github.com/lobehub/lobe-chat/issues/2753) ([6f5be5d](https://github.com/lobehub/lobe-chat/commit/6f5be5d))

</details>

<div align="right">

[![](https://img.shields.io/badge/-BACK_TO_TOP-151515?style=flat-square)](#readme-top)

</div>

xingpingcn commented:
这个新特性是不是在文档中没有更新？

> 客户端数据库db的模式

这个又应该如何配置？

ZeroDeng01 commented:
CLERK_WEBHOOK_SECRET这个配置后是否需要去clerk配置Endpoint？我clerk的相关密钥已经做了配置，用户也可以登录，但是我查看数据库并没有新建clerk的用户信息，现在我的项目完全无法使用。
![image](https://github.com/lobehub/lobe-chat/assets/19732560/036b6e39-4e1c-4976-bdee-0da1d59a5fb8)
我查询日志，就是因为在数据库获取不到用户信息。应为users表为空：
![image](https://github.com/lobehub/lobe-chat/assets/19732560/bf0a9e11-34bc-4db5-a8d9-6a800bad01b9)
![image](https://github.com/lobehub/lobe-chat/assets/19732560/3c885290-1157-4d2a-938d-f1826f2d14ef)
@arvinxx

arvinxx commented:
> CLERK_WEBHOOK_SECRET这个配置后是否需要去clerk配置Endpoint？我clerk的相关密钥已经做了配置，用户也可以登录，但是我查看数据库并没有新建clerk的用户信息，现在我的项目完全无法使用。

需要的，类似这样：

![image](https://github.com/lobehub/lobe-chat/assets/28616219/128d1756-e523-423e-aff4-4bd7fe25e5d2)


目前文档还没完善，有问题可以在 Discord 里咨询

cmj2002 commented:
> 以及有待 @cy948 帮忙实现 next-auth 相关的 db 鉴权方案。

该部分工作是否有相关联的 issue/discussion/pull request ?

arvinxx commented:
> 该部分工作是否有相关联的 issue/discussion/pull request ?

@cy948 感觉可以先写一篇 RFC 来讨论下？

cy948 commented:
@arvinxx 我试完第一个方案之后写个rfc描述下现在的情况吧

gaarry commented:
用上数据库了，朋友也注册我的网站，体验很棒～

gaspire commented:
> 用上数据库了，朋友也注册我的网站，体验很棒～

分享下 教程，谢谢！
